_posts/2021-08-31-modifier-type-clef-primaire-non-referencee-vers-biginit.md

```diff
@@ -0,0 +1,257 @@
+
+L'exposition de la nouvelle colonne, lors du déploiement, a pris un temps négligeable pour les utilisateurs finaux (< 1 seconde).
+
+```shell
+2021-08-27 14:10:40.368443479 +0200 CEST [postdeploy-2719] > pix-api@3.92.0 postdeploy /app
+2021-08-27 14:10:40.368393446 +0200 CEST [postdeploy-2719]
+2021-08-27 14:10:41.566685265 +0200 CEST [postdeploy-2719] Batch 188 run: 1 migrations
+2021-08-27 14:10:42.176473412 +0200 CEST [manager] container [postdeploy-2719] (6128d63ffcff1013ac9e57c1) has stopped
+```
+
+### Mais
+
+#### Un impact non négligeable du traitement de migration sur le temps de réponse
+
+Bien que l'opération soit un succès, lors des étapes préparatoires :
+
+* le temps de réponse de l'API a augmenté d'un facteur 10, en passant d'une moyenne de 30 ms à 300 ms ;
+* lors de l'étape de migration des données, ont eu lieu 4 pics à 1 seconde, d'une demi-heure chacun ;
+* lors de l'étape de validation des contraintes, a eu lieu 1 pic à 1 seconde, d'une heure et demie.
+
+![image info](/assets/images/posts/change-pk-unreferenced-bigint/average-response-time.png)
+
+Malgré cela, l'impact utilisateur a été faible, car :
+
+- cela eu lieu en été, période de faible fréquentation de la plateforme ;
+- le plus long pic a eu lieu à un horaire de faible fréquentation (nuit).
+
+Si l'impact utilisateur s'était avéré plus fort, nous aurions ralenti le traitement en réduisant :
+
+- la taille des batchs ;
+- le rythme d'exécution (pauses périodiques).
+
+#### Un impact sur l'organisation des données sur le disque
+
+Nous avons constaté que la table est bien plus fragmentée après migration qu'avant migration. Bien que nous n'ayons pas fait de test comparatif, il est possible que cela soit dû à notre solution.
+
+En effet :
+
+- la solution native réécrit la table en une seule fois ;
+- notre solution la réécrit en plusieurs fois.
+
+Cela peut avoir deux effets :
+
+- occupation de disque plus importante ;
+- impact sur les temps de réponse.
+
+Nous monitorons actuellement les temps de réponse pour déterminer s'il y a un impact.
+
+## Perspectives
+
+Nous avons prévu de changer le type d'identifiant sur d'autres tables.
+Nous ne pourrons pas appliquer exactement la même approche à l'avenir, pour deux raisons qui seront détaillées. Aussi, nous allons continuer à y travailler et cela fera l'objet d'un nouvel article.
+
+### Diminuer les impacts possibles
+
+Le contexte de faible fréquentation de la plateforme ne peut pas être pris comme un pré-requis. En particulier, l'existence d'un verrou exclusif sur un période de plusieurs minutes est problématique.
+
+Nous pensons à :
+
+- modifier notre approche en passant d'une colonne temporaire à une table temporaire ;
+- rendre le traitement de migration paramétrable à chaud, en passant de variables d'environnement à une table de paramétrage.
+
+### Généraliser la solution
+
+La table qui a été migrée n'était pas référencée par d'autres tables.
+Cela était implicite, car elle était la plus volumineuse.
```

Comment:
Hmm… tu veux dire que si une table est plus volumineuse (en nombre de lignes) que toutes les autres, elle ne peut pas être référencée par d'autres tables ? Une sorte de [Principe des tiroirs](https://fr.wikipedia.org/wiki/Principe_des_tiroirs) à l'envers ? Ça serait intéressant à défendre 😉 mais je ne suis pas sûr que ce soit clair pour le lecteur ici.

Reply:
Il y a effectivement (diable, ce Jonathan voit clair dans mon jeu !) des hypothèses implicites:
- la FK est NOT NULL
- la majorité des valeurs est référencée au moins une fois.


Ca fait beaucoup d'implicité, je l'ai enlevé.